editor/src/components/editor/canvas-toolbar.tsx

```diff
@@ -0,0 +1,344 @@
+    (store) => store.editor.floatingInsertMenu.insertMenuMode,
+    'CanvasToolbar insertMenuMode',
+  )
+
+  const openFloatingInsertMenuCallback = React.useCallback(() => {
+    dispatch([
+      openFloatingInsertMenu({
+        insertMenuMode: 'insert',
+        parentPath: null,
+        indexPosition: null,
+      }),
+    ])
+  }, [dispatch])
+  const openFloatingConvertMenuCallback = React.useCallback(() => {
+    dispatch([
+      openFloatingInsertMenu({
+        insertMenuMode: 'convert',
+      }),
+    ])
+  }, [dispatch])
+  const openFloatingWrapInMenuCallback = React.useCallback(() => {
+    dispatch([
+      openFloatingInsertMenu({
+        insertMenuMode: 'wrap',
+      }),
+    ])
+  }, [dispatch])
+
+  const wrapInDivCallback = React.useCallback(() => {
+    dispatch([wrapInView(selectedViewsRef.current, 'default-empty-div')])
+  }, [dispatch, selectedViewsRef])
+
+  const insertMenuSelected = useEditorState(
+    (store) => store.editor.rightMenu.selectedTab === RightMenuTab.Insert,
+    'CanvasToolbar insertMenuSelected',
+  )
+
+  const selectInsertMenuPane = React.useCallback(() => {
+    let actions: Array<EditorAction> = []
+    if (!insertMenuSelected) {
+      actions.push(setPanelVisibility('rightmenu', true))
+    }
+    actions.push(setRightMenuTab(RightMenuTab.Insert))
+    dispatch(actions)
+  }, [dispatch, insertMenuSelected])
+
+  const zoomLevel = useEditorState((store) => store.editor.canvas.scale, 'CanvasToolbar zoomLevel')
+
+  const zoomIn = React.useCallback(
+    () => dispatch([CanvasActions.zoom(Utils.increaseScale(zoomLevel))]),
+    [dispatch, zoomLevel],
+  )
+  const zoomOut = React.useCallback(
+    () => dispatch([CanvasActions.zoom(Utils.decreaseScale(zoomLevel))]),
+    [dispatch, zoomLevel],
+  )
+
+  const isLiveMode = useEditorState(
+    (store) => store.editor.mode.type === 'live',
+    'TopMenu isLiveMode',
+  )
+  const toggleLiveMode = React.useCallback(() => {
+    if (isLiveMode) {
+      dispatch([switchEditorMode(EditorModes.selectMode())])
+    } else {
+      dispatch([switchEditorMode(EditorModes.liveMode())])
+    }
+  }, [dispatch, isLiveMode])
+
+  const resetCanvasCallback = React.useCallback(() => {
+    dispatch([resetCanvas()])
+  }, [dispatch])
+
+  const isCodeEditorVisible = useEditorState(
+    (store) => store.editor.interfaceDesigner.codePaneVisible,
+    'CanvasToolbar isCodeEditorVisible',
+  )
+  const toggleCodeEditorVisible = React.useCallback(
+    () => dispatch([setPanelVisibility('codeEditor', !isCodeEditorVisible)]),
+    [dispatch, isCodeEditorVisible],
+  )
+
+  return (
+    <FlexColumn
+      style={{
+        position: 'absolute',
+        top: topMenuHeight + 8,
+        left: effectiveNavigatorWidth + 8,
+        alignItems: 'stretch',
+        width: 64,
+        borderRadius: 4,
+        backgroundColor: theme.bg0.value,
+        boxShadow: UtopiaStyles.popup.boxShadow,
+      }}
+    >
+      <FlexColumn style={{ padding: 4 }}>
```

Comment:
I think I should introduce more code organization in a follow-up PR, splitting this big chunk of JSX into sub-components.

Reply:
Ha! Yes please :D